src/sql/workbench/contrib/notebook/browser/cellToolbarActions.ts

```diff
@@ -0,0 +1,114 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Action } from 'vs/base/common/actions';
+import { localize } from 'vs/nls';
+import { INotebookService, INotebookEditor } from 'sql/workbench/services/notebook/browser/notebookService';
+import { NotebookComponent } from 'sql/workbench/contrib/notebook/browser/notebook.component';
+import { ICellModel } from 'sql/workbench/services/notebook/browser/models/modelInterfaces';
+import { CellType } from 'sql/workbench/services/notebook/common/contracts';
+import { ToggleableAction } from 'sql/workbench/contrib/notebook/browser/notebookActions';
+
+export class CellToolbarAction extends Action {
+
+	constructor(
+		id: string,
+		label: string,
+		cssClass: string,
+		tooltip: string,
+		private _cellModel: ICellModel,
+		@INotebookService private _notebookService: INotebookService
+	) {
+		super(id, label, cssClass);
+		this._tooltip = tooltip;
+	}
+	// Todo: Add actions references here?
+}
+
+export class AddCellAction extends Action {
+	public cellType: CellType;
+
+	constructor(
+		id: string, label: string, cssClass: string
+	) {
+		super(id, label, cssClass);
+	}
+	public async run(context: INotebookEditor): Promise<any> {
+		//Add Cell after current selected cell.
+		let index = 0;
+		if (context && context.cells) {
+			let notebookcomponent = context as NotebookComponent;
+			let id = notebookcomponent.activeCellId;
+			if (id) {
+				index = context.cells.findIndex(cell => cell.id === id);
+				index = index + 1;
+			}
+		}
+		context.addCell(this.cellType, index);
+	}
+}
+
+/**
+ * Not yet implemeneted
+ * Goal: show more actions available depending on what cell the toolbar is for.
+ * (ie. runAlActions when on code cell.)
+ */
+export class MoreActions extends Action {
```

Comment:
Corresponds to the moreActionsButton I had wired up in the cell toolbar component. Again, I don't know how to leverage the cellToggleMoreActions here.

Reply:
The cells have an ActionBar that they populate with the actions

https://github.com/microsoft/azuredatastudio/blob/master/src/sql/workbench/contrib/notebook/browser/cellToggleMoreActions.ts#L58

Looking at CelltoggleMoreActions it seems like you'd just create a host element and then create the CellToggleMoreActions with that https://github.com/microsoft/azuredatastudio/blob/master/src/sql/workbench/contrib/notebook/browser/cellViews/textCell.component.ts#L146